### ISSUE TITLE
Metrics of ignored/initialized pods in a scheduler performance test are benchmarked

### ISSUE THREAD

**Issue by Huang-Wei** [kubernetes/kubernetes#99218]
#### What happened:

In the new scheduler performance testing framework, `collectMetrics`(bool) is provided to benchmark performance of pods created in an "operator": (L11 below)

https://github.com/kubernetes/kubernetes/blob/df8dee80c838bfb50532d1509c1a77a9c4463a11/test/integration/scheduler_perf/config/performance-config.yaml#L1-L17

However, the initialized Pods created in L5~L7 seem to be also benchmarked, although it doesn't specify `collectMetrics: true`. (please see the reproduce steps below).

I think it's b/c once scheduler starts, the associated Prometheus artifacts are also up, and it's a single in-memory datastore. So no matter we start to collect the metric in which step, the prometheus data is already up-to-date. In the above workload config, the final data have benchmarked 1500 pods instead of 1000 pods. I guess it won't work either if running `hist.Clear()` in the pod initialization step or not. Unless we manage a breakpoint at the end of each `createPods` step, and calculate delta histogram for the step that specifies `collectMetrics: true`.

#### What you expected to happen:

Pods created in step `createPods` without `collectMetrics: true` are not benchmarked.

#### How to reproduce it (as minimally and precisely as possible):

Apply the following diff:

```diff
diff --git a/staging/src/k8s.io/component-base/metrics/testutil/metrics.go b/staging/src/k8s.io/component-base/metrics/testutil/metrics.go
index ac66772c446..c9319c769e2 100644
--- a/staging/src/k8s.io/component-base/metrics/testutil/metrics.go
+++ b/staging/src/k8s.io/component-base/metrics/testutil/metrics.go
@@ -273,6 +273,10 @@ func (hist *Histogram) Quantile(q float64) float64 {
 		})
 	}
 
+	for _, b := range buckets {
+		fmt.Printf("[DEBUG] %v le %v\n", b.count, b.upperBound)
+	}
+
 	return bucketQuantile(q, buckets)
 }
 
diff --git a/test/integration/scheduler_perf/scheduler_perf_test.go b/test/integration/scheduler_perf/scheduler_perf_test.go
index 3ed259e8d01..beba1b7d069 100644
--- a/test/integration/scheduler_perf/scheduler_perf_test.go
+++ b/test/integration/scheduler_perf/scheduler_perf_test.go
@@ -49,9 +49,9 @@ const (
 var (
 	defaultMetricsCollectorConfig = metricsCollectorConfig{
 		Metrics: []string{
-			"scheduler_scheduling_algorithm_predicate_evaluation_seconds",
-			"scheduler_scheduling_algorithm_priority_evaluation_seconds",
-			"scheduler_e2e_scheduling_duration_seconds",
+			// "scheduler_scheduling_algorithm_predicate_evaluation_seconds",
+			// "scheduler_scheduling_algorithm_priority_evaluation_seconds",
+			// "scheduler_e2e_scheduling_duration_seconds",
 			"scheduler_pod_scheduling_duration_seconds",
 		},
 	}
```

And then run:

```
make test-integration WHAT=./test/integration/scheduler_perf KUBE_TEST_VMODULE="''" KUBE_TEST_ARGS="-alsologtostderr=false -logtostderr=false -run=^$$ -benchtime=1ns -bench=BenchmarkPerfScheduling/SchedulingBasic/500Nodes"
```

You will get an output sth. like:

```
+++ [0218 15:31:02] Running tests without code coverage
[DEBUG] 263 le 0.01
[DEBUG] 367 le 0.02
[DEBUG] 449 le 0.04
[DEBUG] 586 le 0.08
[DEBUG] 865 le 0.16
[DEBUG] 1210 le 0.32
[DEBUG] 1213 le 0.64
[DEBUG] 1446 le 1.28
[DEBUG] 1495 le 2.56
[DEBUG] 1500 le 5.12
[DEBUG] 1500 le 10.24
[DEBUG] 1500 le 20.48
[DEBUG] 1500 le 40.96
[DEBUG] 1500 le 81.92
[DEBUG] 1500 le 163.84
[DEBUG] 1500 le 327.68
[DEBUG] 1500 le 655.36
[DEBUG] 1500 le 1310.72
[DEBUG] 1500 le 2621.44
[DEBUG] 1500 le 5242.88
[DEBUG] 1500 le +Inf
......
```

/sig scheduling

**Comment by Huang-Wei** [comment]
@ingvagabund Could you please help confirm if it's an issue?

**Comment by Huang-Wei** [comment]
cc @alculquicondor @adtac @ahg-g

**Comment by ingvagabund** [comment]
Unfortunately, `collectMetrics` does not apply for metrics gathered from `legacyregistry.DefaultGatherer`.

> I think it's b/c once scheduler starts, the associated Prometheus artifacts are also up, and it's a single in-memory datastore.

As you mentioned. The default gatherer collects all metrics produced by the scheduler. It's the scheduler itself who knows pod names and would have to decide which pods will get ignored when updating metrics. The question is if computing metrics for a subset of pods processed by the scheduler is a valid use case.

**Comment by alculquicondor** [comment]
> Unless we manage a breakpoint at the end of each createPods step, and calculate delta histogram for the step that specifies collectMetrics: true.

This is probably what we want. Is something like that possible @ingvagabund ?

**Comment by alculquicondor** [comment]
I think being able to absolutely ignore certain pods for metrics is an overkill.

**Comment by Huang-Wei** [comment]
> The question is if computing metrics for a subset of pods processed by the scheduler is a valid use case.

The use case is to ignore the pods created in steps that don't specify `collectMetrics: true`, so as to collect the Pod metrics create at particular step(s) (`collectMetrics: true`). Otherwise, the semantics of `collectMetrics` option is quite misleading.

> I think being able to absolutely ignore certain pods for metrics is an overkill.

Nope, that's not the intention. We don't need to target certain pods, like using label selector or whatever. Just as I mentioned, measure pods created at particular steps.

**Comment by alculquicondor** [comment]
I would do something like the following:

if a step has collectMetrics:false, we wait for them to finish scheduling, and then we take a snapshot of the histogram. Then for the next step that has collectMetrics:true, we take the diff with the latest snapshot to report the results. This might not be perfect, as we don't control if the metrics have finished syncing when we take the snapshot. But I think that's fine. Would doing this be possible?

Now, if we have steps that run in parallel, one with collectMetrics:false and another with collectMetrics:true, then this doesn't work.

Note that I don't know the details of how the gatherer works.

**Comment by Huang-Wei** [comment]
> if a step has collectMetrics:false, we wait for them to finish scheduling, and then we take a snapshot of the histogram. Then for the next step that has collectMetrics:true, we take the diff with the latest snapshot to report the results.

That adheres to what I have in my mind - we can only manage this on client side.

> This might not be perfect, as we don't control if the metrics have finished syncing when we take the snapshot. But I think that's fine. Would doing this be possible?

I think it's fine. As we are already fetching the "snapshot" when pods get scheduled when `collectMetrics` equals true.

> Now, if we have steps that run in parallel, one with collectMetrics:false and another with collectMetrics:true, then this doesn't work.

True.

An alternative is to design the case to be like: the step with `collectMetrics:false` to employ Pods that are _all unschedulable_, and then specify a custom metric "scheduler_e2e_scheduling_duration_seconds[result="unschedulable"]" to target a **particular** histogram instead of **all** histograms (default option) inside a histogramVec.
Note this can only be possible when #99228 and some subsequent changes get in place.

**Comment by ingvagabund** [comment]
We might just call Histogram.Clear() method which will clear what has been collected so far at the end of [createPodsOp](https://github.com/kubernetes/kubernetes/blob/a42ff247afeb2f2d7aaf1d28a7cb171d722fb527/test/integration/scheduler_perf/scheduler_perf_test.go#L378) phase:
```
if !concreteOp.CollectMetrics {
	hist, err := testutil.GetHistogramFromGatherer(legacyregistry.DefaultGatherer, metric)
	...
	hist.Clear()
}
```

**Comment by alculquicondor** [comment]
WDYT @adtac?

**Comment by Huang-Wei** [comment]
@ingvagabund I'm afraid `Clear()` doesn't work. That was my initial thought as well, but no luck. I guess it's b/c, theoretically, we cannot reset a metric on the server side - Promoetheus' model doesn't support this. But let me know if I miss anything.

**Comment by ingvagabund** [comment]
The performance tests assume all the metrics all gathered and accessed in the same binary. As long as that's the case, `Clear()` works

**Comment by Huang-Wei** [comment]
> The performance tests assume all the metrics all gathered and accessed in the same binary. As long as that's the case, `Clear()` works. We already call `Clear()` in the code:

That was my understanding as well, but no luck :(

Here is a diff to call `Clear()` in case `createNodesOp` when "concreteOp.CollectMetrics" is false:

<details><summary>Diff:</summary>
<p>

```diff
diff --git a/staging/src/k8s.io/component-base/metrics/testutil/metrics.go b/staging/src/k8s.io/component-base/metrics/testutil/metrics.go
index ac66772c446..c9319c769e2 100644
--- a/staging/src/k8s.io/component-base/metrics/testutil/metrics.go
+++ b/staging/src/k8s.io/component-base/metrics/testutil/metrics.go
@@ -273,6 +273,10 @@ func (hist *Histogram) Quantile(q float64) float64 {
 		})
 	}
 
+	for _, b := range buckets {
+		fmt.Printf("[DEBUG] %v le %v\n", b.count, b.upperBound)
+	}
+
 	return bucketQuantile(q, buckets)
 }
 
diff --git a/test/integration/scheduler_perf/scheduler_perf_test.go b/test/integration/scheduler_perf/scheduler_perf_test.go
index 3ed259e8d01..8fe9ec23482 100644
--- a/test/integration/scheduler_perf/scheduler_perf_test.go
+++ b/test/integration/scheduler_perf/scheduler_perf_test.go
@@ -49,9 +49,9 @@ const (
 var (
 	defaultMetricsCollectorConfig = metricsCollectorConfig{
 		Metrics: []string{
-			"scheduler_scheduling_algorithm_predicate_evaluation_seconds",
-			"scheduler_scheduling_algorithm_priority_evaluation_seconds",
-			"scheduler_e2e_scheduling_duration_seconds",
+			// "scheduler_scheduling_algorithm_predicate_evaluation_seconds",
+			// "scheduler_scheduling_algorithm_priority_evaluation_seconds",
+			// "scheduler_e2e_scheduling_duration_seconds",
 			"scheduler_pod_scheduling_duration_seconds",
 		},
 	}
@@ -348,13 +348,12 @@ func runWorkload(b *testing.B, tc *testCase, w *workload) []DataItem {
 			} else {
 				namespace = fmt.Sprintf("namespace-%d", opIndex)
 			}
-			var collectors []testDataCollector
+			collectors := getTestDataCollectors(podInformer, fmt.Sprintf("%s/%s", b.Name(), namespace), namespace, tc.MetricsCollectorConfig)
 			var collectorCtx context.Context
 			var collectorCancel func()
 			if concreteOp.CollectMetrics {
 				collectorCtx, collectorCancel = context.WithCancel(ctx)
 				defer collectorCancel()
-				collectors = getTestDataCollectors(podInformer, fmt.Sprintf("%s/%s", b.Name(), namespace), namespace, tc.MetricsCollectorConfig)
 				for _, collector := range collectors {
 					go collector.run(collectorCtx)
 				}
@@ -385,6 +384,10 @@ func runWorkload(b *testing.B, tc *testCase, w *workload) []DataItem {
 					dataItems = append(dataItems, collector.collect()...)
 				}
 				mu.Unlock()
+			} else {
+				for _, collector := range collectors {
+					collector.clear()
+				}
 			}
 
 		case *barrierOp:
@@ -418,6 +421,7 @@ func runWorkload(b *testing.B, tc *testCase, w *workload) []DataItem {
 type testDataCollector interface {
 	run(ctx context.Context)
 	collect() []DataItem
+	clear()
 }
 
 func getTestDataCollectors(podInformer coreinformers.PodInformer, name, namespace string, mcc *metricsCollectorConfig) []testDataCollector {
diff --git a/test/integration/scheduler_perf/util.go b/test/integration/scheduler_perf/util.go
index 7b846c9ed8f..2f03d4eed01 100644
--- a/test/integration/scheduler_perf/util.go
+++ b/test/integration/scheduler_perf/util.go
@@ -179,6 +179,21 @@ func (pc *metricsCollector) collect() []DataItem {
 	return dataItems
 }
 
+func (pc *metricsCollector) clear() {
+	legacyregistry.Reset()
+	for _, metric := range pc.Metrics {
+		hist, err := testutil.GetHistogramFromGatherer(legacyregistry.DefaultGatherer, metric)
+		if err != nil {
+			klog.Error(err)
+			continue
+		}
+		fmt.Printf("[DEBUG] %v's histogram get cleared\n", metric)
+		hist.Clear()
+		// This doesn't work either.
+		// hist.Reset()
+	}
+}
+
 func collectHistogram(metric string, labels map[string]string) *DataItem {
 	hist, err := testutil.GetHistogramFromGatherer(legacyregistry.DefaultGatherer, metric)
 	if err != nil {
@@ -287,3 +302,5 @@ func (tc *throughputCollector) collect() []DataItem {
 
 	return []DataItem{throughputSummary}
 }
+
+func (tc *throughputCollector) clear() {}

``` 

</p>
</details>

And below is the output - it's still 1500 (500 init pods plus 1000 measured pods) instead of 1000 being benchmarked:

<details><summary>Output:</summary>
<p>

```
[DEBUG] scheduler_pod_scheduling_duration_seconds's histogram get cleared
[DEBUG] 127 le 0.01
[DEBUG] 219 le 0.02
[DEBUG] 392 le 0.04
[DEBUG] 741 le 0.08
[DEBUG] 1219 le 0.16
[DEBUG] 1285 le 0.32
[DEBUG] 1285 le 0.64
[DEBUG] 1470 le 1.28
[DEBUG] 1494 le 2.56
[DEBUG] 1500 le 5.12
[DEBUG] 1500 le 10.24
[DEBUG] 1500 le 20.48
[DEBUG] 1500 le 40.96
[DEBUG] 1500 le 81.92
[DEBUG] 1500 le 163.84
[DEBUG] 1500 le 327.68
[DEBUG] 1500 le 655.36
[DEBUG] 1500 le 1310.72
[DEBUG] 1500 le 2621.44
[DEBUG] 1500 le 5242.88
[DEBUG] 1500 le +Inf
```

</p>
</details>

**Comment by Huang-Wei** [comment]
But I do find a way to fix this: #99472.
